libsolidity/analysis/ConstantEvaluator.cpp

```diff
@@ -57,6 +58,21 @@ void ConstantEvaluator::endVisit(BinaryOperation const& _operation)
 				" and " +
 				right->toString()
 			);
+
+		if (auto const rationalCommonType = dynamic_cast<RationalNumberType const*>(commonType))
+		{
+			auto leftType = type(_operation.leftExpression()).type;
+			auto rightType = type(_operation.rightExpression()).type;
+			if (leftType && leftType->category() == Type::Category::Integer &&
```

Comment:
If left and right are integers, then comon should also be integer. How can thiscondition be true?

Reply:
Because `RationalNumberType::binaryOperatorResult(Token _op, TypePointer _rhs)` does return a `RationalNumberType` regardless of the type of `_rhs`. Internally every literal is stored as `RationalNumberType`.

https://github.com/ethereum/solidity/blob/develop/libsolidity/ast/Types.cpp#L1208

Reply:
The linked function has many early returns.